File: java/Exercises.java
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Exercises {
    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }

    static Optional<String> firstThenLowerCase(List<String> a, Predicate<String> p) {
        return a.stream()
                    .filter(p)
                    .findFirst()
                    .map(x -> x.toLowerCase());
    }

    public static sayer say(){
        return new sayer("");
    }

    public static sayer say(String phrase){
        return new sayer(phrase);
    }

    public static record sayer(String phrase){
        public sayer and(String message){
            return new sayer(phrase + " " + message);
        }
    }

    public static int meaningfulLineCount(String filename) throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            return (int) reader.lines()
                    .filter(line -> !line.trim().isEmpty())
                    .filter(line -> line.trim().charAt(0) != '#')
                    .count();
        } catch (IOException e){
            throw e;
        }
    }
}

record Quaternion(double a, double b, double c, double d) {
    public final static Quaternion ZERO = new Quaternion(0.0, 0.0, 0.0, 0.0);
    public final static Quaternion I = new Quaternion(0.0, 1.0, 0.0, 0.0);
    public final static Quaternion J = new Quaternion(0.0, 0.0, 1.0, 0.0);
    public final static Quaternion K = new Quaternion(0.0, 0.0, 0.0, 1.0);

    public Quaternion {
        if (Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d)) {
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }

    Quaternion plus(Quaternion other) {
        return new Quaternion(a + other.a(), b + other.b(), c + other.c(), d + other.d());
    }

    Quaternion times(Quaternion other) {
        return new Quaternion(a * other.a() - b * other.b() - c * other.c() - d * other.d(),
                          b * other.a() + a * other.b() + c * other.d() - d * other.c(),
                          a * other.c() - b * other.d() + c * other.a() + d * other.b(),
                          a * other.d() + b * other.c() - c * other.b() + d * other.a());
    }

    List<Double> coefficients() {
        return List.copyOf(List.of(a, b, c, d));
    }

    Quaternion conjugate() {
        return new Quaternion(a, -b, -c, -d);
    }

    private String plusOrMinus(Double coeff) {
        if (Math.abs(coeff) == 1.0) return ((coeff > 0) ? ("+") : ("-"));
        return ((coeff > 0) ? ("+") : (""));
    }

    @Override
    public String toString() {
        List<Double> coeffs = this.coefficients();
        List<String> dims = List.of("","i","j","k");
        String coeffStr = "";
        int zeroCounter = 0;
        for (int i = 0; i < coeffs.size(); i++){
            if (coeffs.get(i) == 0) {
                zeroCounter += 1;
                continue;
            } else if (i == 0) {
                coeffStr += Double.toString(coeffs.get(i));
            } else if (Math.abs(coeffs.get(i)) == 1.0) {
                coeffStr += (plusOrMinus(coeffs.get(i)) + dims.get(i));
            } else {
                coeffStr += (plusOrMinus(coeffs.get(i)) + Double.toString(coeffs.get(i)) + dims.get(i));
            }
        }
        if (zeroCounter == 4) {
            return "0";
        } else {
            if (coeffStr.charAt(0) == '+') {
                coeffStr = coeffStr.substring(1);
            }
            return coeffStr;
        }
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Quaternion)) {
            return false;
        }
        Quaternion other = (Quaternion) o;
        return ((a == other.a()) && (b == other.b()) && (c == other.c()) && (d == other.d()));
    }
}

sealed interface BinarySearchTree permits Empty, Node {
    int size();
    boolean contains(String value);
    BinarySearchTree insert(String value);
}

final record Empty() implements BinarySearchTree {
    @Override
    public int size() {
        return 0;
    }
    @Override
    public boolean contains(String item) {
        return false;
    }
    @Override
    public BinarySearchTree insert(String value) {
        return new Node(value, this, this);
    }

    @Override
    public String toString() {
        return "()";
    }
}

final class Node implements BinarySearchTree {
    private final String value;
    private final BinarySearchTree left;
    private final BinarySearchTree right;
    
    Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }
    
    @Override
    public int size() {
        return 1 + this.right.size() + this.left.size();
    }

    @Override
    public boolean contains(String value) {
        return this.value.equals(value) || this.left.contains(value) || this.right.contains(value);
    }

    @Override
    public BinarySearchTree insert(String newVal) {
        if (newVal.compareTo(this.value) < 0) {
            return new Node(this.value, this.left.insert(newVal), this.right);
        } else if (newVal.compareTo(this.value) > 0) {
            return new Node(this.value, this.left, this.right.insert(newVal));
        } else {
            return this;
        }
    }

    private String emptyOrNot(BinarySearchTree node) {
        return (node instanceof Empty) ? "" : node.toString();
    }

    @Override
    public String toString() {
        return "(" + emptyOrNot(left) + value + emptyOrNot(right) + ")";
    }
}

File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.File
import java.io.IOException
import kotlin.math.abs

fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }

    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

fun firstThenLowerCase(of: List<String>, predicate: (String) -> Boolean): String? {
    val filtered: List<String> = of.filter(predicate)
    return (if (!filtered.isEmpty()) filtered[0] else null)?.lowercase()
}

class say (new_phrase: String = "") {
    val phrase: String = new_phrase

    fun and(next_phrase: String): say {
        return say(phrase + " " + next_phrase)
    }
}

@Throws(IOException::class)
fun meaningfulLineCount(filename: String): Long {
File(filename).bufferedReader().use {
    return it.readLines().map { 
        it.trim() 
    }.filter{ 
        !it.isNullOrBlank() 
    }.filter { 
        it.elementAtOrNull(0) != '#'
    }.count().toLong()
    }
}

data class Quaternion(val a: Double, val b: Double, val c: Double, val d: Double) {
    // This takes the place of static member variables
    companion object {
        val ZERO = Quaternion(0.0,0.0,0.0,0.0)
        val I = Quaternion(0.0,1.0,0.0,0.0)
        val J = Quaternion(0.0,0.0,1.0,0.0)
        val K = Quaternion(0.0,0.0,0.0,1.0)
    }

    init {
        require(!a.isNaN() && !b.isNaN() && !c.isNaN() && !d.isNaN()) {
            "Coordinates can not be NaN"
        }
    }

    operator fun plus(other: Quaternion): Quaternion {
        return Quaternion(this.a + other.a, this.b + other.b, this.c + other.c, this.d + other.d)
    }

    operator fun times(other: Quaternion): Quaternion {
        val mul_a: Double = this.a * other.a - this.b * other.b - this.c * other.c - this.d * other.d
        val mul_b: Double = this.a * other.b + this.b * other.a + this.c * other.d - this.d * other.c
        val mul_c: Double = this.a * other.c - this.b * other.d + this.c * other.a + this.d * other.b
        val mul_d: Double = this.a * other.d + this.b * other.c - this.c * other.b + this.d * other.a
        return Quaternion(mul_a, mul_b, mul_c, mul_d)
    }

    fun coefficients(): List<Double> {
        return listOf(this.a, this.b, this.c, this.d)
    }

    fun conjugate(): Quaternion {
        return Quaternion(this.a, -this.b, -this.c, -this.d)
    }

    override fun toString(): String {
        fun coefToString(coef: Double, comp: String): String {
            val sign: String = if (coef >= 0) "+" else "-"
            val scalar: String = if (abs(coef) != 1.0 || comp.isEmpty()) "${abs(coef)}" else ""
            return "$sign$scalar$comp"
        }
        
        val components: List<Pair<Double, String>> = listOf(Pair(this.a, ""), Pair(this.b, "i"), Pair(this.c, "j"), Pair(this.d, "k"))
        val coefStrings: List<String> = components.filter {
            (coef, _): Pair<Double, String> -> coef != 0.0
        }.map {
            (coef, comp): Pair<Double, String> -> coefToString(coef, comp)
        }
        if (coefStrings.isEmpty()) {
            return "0"
        }
        else {
            val coefString: String = coefStrings.joinToString(separator="")
            return if (coefString[0] == '+') coefString.substring(1) else coefString
        }
    }
}

sealed interface BinarySearchTree {
    fun size(): Int
    fun contains(value: String): Boolean
    fun insert(value: String): BinarySearchTree

    object Empty : BinarySearchTree {
        override fun size(): Int = 0
        override fun contains(value: String): Boolean = false
        override fun insert(value: String): BinarySearchTree = Node(value, Empty, Empty)
        override fun toString(): String = "()"
    }

    data class Node(
        val value: String,
        val left: BinarySearchTree = Empty,
        val right: BinarySearchTree = Empty
    ) : BinarySearchTree {
        override fun size(): Int = 1 + left.size() + right.size()

        override fun contains(value: String): Boolean {
            return when {
                value == this.value -> true
                value < this.value -> left.contains(value)
                else -> right.contains(value)
            }
        }

        override fun insert(newVal: String): BinarySearchTree {
            return when {
                newVal == this.value -> this
              newVal < this.value -> Node(this.value, left.insert(newVal), right)
                else -> Node(this.value, left, right.insert(newVal))
            }
        }

        override fun toString(): String {
            val leftStr = if (left == Empty) "" else left.toString()
            val rightStr = if (right == Empty) "" else right.toString()
            return "($leftStr$value$rightStr)"
        }
    }
}
